File: 07_Java/Homework/Chapter_11-13/HomeWork/src/Test5.java
public class Test5 {
    public static void main(String... args) {
        C4 o1 = new D4();
        o1.m(1, 1);            //①
        o1.m(1.0, 1.0);            //②
        o1.m(1.0f, 1.0f);        //③

        D4 o2 = new D4();
        o2.m(1, 1);            //④
        o2.m(1.0, 1.0);            //⑤
        o2.m(1.0f, 1.0f);        //⑥
    }
}

class C4 {
    public void m(int x, int y) {
        System.out.println("C's m(int,int)");
    }

    public void m(double x, double y) {
        System.out.println("C' m(double,double)");
    }
}

class D4 extends C4 {
    public void m(float x, float y) {
        System.out.println("D's m(float,float)");
    }

    public void m(int x, int y) {
        System.out.println("D's m(int,int)");
    }
}

File: 08_Assembly Language/Source Code/C12 中断和异常/test5/test5.cpp
// test5.cpp : 定义控制台应用程序的入口点。
//

#include "stdafx.h"
#include "stdio.h"
#include "windows.h"
#include <winternl.h>

int exception_filter(LPEXCEPTION_POINTERS p_exinfo)
{
	printf("Error address %x \n", p_exinfo->ExceptionRecord->ExceptionAddress);
	printf("Error Code %08x \n", p_exinfo->ExceptionRecord->ExceptionCode);
	printf("CPU register: \n");
	printf("eax= %08x ebx=%08x  ecx=%08x  edx=%08x \n", p_exinfo->ContextRecord->Eax,
		p_exinfo->ContextRecord->Ebx, p_exinfo->ContextRecord->Ecx,
		p_exinfo->ContextRecord->Edx);
	if (p_exinfo->ExceptionRecord->ExceptionCode == EXCEPTION_ACCESS_VIOLATION)
	{
		puts("存储保护异常");
	}
	if (p_exinfo->ExceptionRecord->ExceptionCode == EXCEPTION_INT_DIVIDE_BY_ZERO)
	{
		puts("被0除异常");
	}
	return 1;
}

void main()
{
	int  x = 10;
	int  y = 0;
	int  result=3;
	int  select;
	TEB  *p;
	puts("hello");
	scanf_s("%d", &select);
	__try
	{
			if (select == 1) {
			  int* p;
			  p = 0;
			  *p = 45;	// 该语句会导致一个异常  EXCEPTION_ACCESS_VIOLATION
			  puts("This Message will not occur ...ACCESS_VIOLATION ");
			}
			if (select==2) {
				result = x / y;  // 该语句会导致一个异常  EXCEPTION_INT_DIVIDE_BY_ZERO
				puts("This Message will not occur ...INT_DIVIDE_BY_ZERO ");
			}
			if (select != 1 && select != 2) {
				puts("no exception occur");
			}
		}
	__except (exception_filter(GetExceptionInformation()))
	   {
			puts("异常已处理");
		}
	puts("world");
	result = x / y;  // 该语句会导致一个异常  EXCEPTION_INT_DIVIDE_BY_ZERO

	getchar();
}


